### ISSUE TITLE
Is split direction random for missing value on training data without missing at all?

### ISSUE THREAD

**Issue by is03wlei** [dmlc/xgboost#10637]
Is the [comparison](https://github.com/dmlc/xgboost/blob/master/src/tree/gpu_hist/evaluate_splits.cu#L31) here going to be random due to precision?

My training data has no missing value at all. But it has different split directions for missing value on two devices.

1. Apply M1: all missing value splits direct to right child.
2. A linux server: 2/3 missing values splits direct to left child.

Is it related with how different GPUs handle computation and their precision? Apple M1 is more deterministic on GPU computation?

XGBOOST_DEVICE float LossChangeMissing(const GradientPairInt64 &scan,
                                       const GradientPairInt64 &missing,
                                       const GradientPairInt64 &parent_sum,
                                       const GPUTrainingParam &param, bst_node_t nidx,
                                       bst_feature_t fidx,
                                       TreeEvaluator::SplitEvaluator<GPUTrainingParam> evaluator,
                                       bool &missing_left_out, const GradientQuantiser& quantiser) {  // NOLINT
  const auto left_sum = scan + missing;
  float missing_left_gain = evaluator.CalcSplitGain(
      param, nidx, fidx, quantiser.ToFloatingPoint(left_sum),
      quantiser.ToFloatingPoint(parent_sum - left_sum));
  float missing_right_gain = evaluator.CalcSplitGain(
      param, nidx, fidx, quantiser.ToFloatingPoint(scan),
      quantiser.ToFloatingPoint(parent_sum - scan));

  missing_left_out = missing_left_gain > missing_right_gain;
  return missing_left_out?missing_left_gain:missing_right_gain;
}

**Comment by trivialfis** [comment]
It's not random, but can be hardware-dependent since we employ different implementations for different hardware. You referenced the GPU implementation (CUDA), which is deterministic as we are actually operating in integer space instead of floating point space. For the CPU implementation, if the data is dense, the backward pass is skipped entirely https://github.com/dmlc/xgboost/blob/77c844cef73b6b81d5c1da4f63b925c83fd98bad/src/tree/hist/evaluate_splits.h#L389 .

**Comment by is03wlei** [comment]
Thanks for replying. I conducted more experiments with different tree methods on the same hardware(linux server).
There is still no missing value in train data.
    * exact: all splits direct to left child
    * hist: all splits direct to right child
    * gpu_hist:
        * splits redirects to right if creating dmatrix from scipy sparse matrix(read from libsvm format data)
                   X, y, _ = load_svmlight_file(file_name, query_id=True, zero_based=True, dtype=np.float32)
                   dmatrix = xgb.DMatrix(X, label=y)
        * 2/3 splits direct to left and 1/3 splits direct to right child if creating dmatrix from libsvm format data directly
                   dmatrix = xgb.DMatrix(file_name)

Could you help to understand why two different loadings have different behaviors for gpu_hist?

**Comment by trivialfis** [comment]
I can't provide a definitety answer without seeing the data. Would be great if you can share a reproducible example.
One guess is scipy converts all zeros to missing.

**Comment by is03wlei** [comment]
This is also one of my guesses. But my data doesn't have zero values on the column. My data is super large and I will provide a small version which can reproduce the issue.

**Comment by trivialfis** [comment]
Synthetic data works as well.

**Comment by is03wlei** [comment]
[train.weight.txt](https://github.com/user-attachments/files/16836048/train.weight.txt)
[train.txt](https://github.com/user-attachments/files/16836051/train.txt)
[train.weight.txt](https://github.com/user-attachments/files/16836056/train.weight.txt)
[reproduce.py.txt](https://github.com/user-attachments/files/16836077/reproduce.py.txt)
[model.xgboost.dump-22.json](https://github.com/user-attachments/files/16836098/model.xgboost.dump-22.json)

PYTHONPATH='./' python3 xgbranking/reproduce.py --train_file train
I uploaded script and training data which can reproduce the issue. Also uploaded the model.
